warehouse/logfield/logfield.go

```diff
@@ -23,7 +25,96 @@ const (
 	LoadFileType               = "loadFileType"
 	ErrorMapping               = "errorMapping"
 	DestinationCredsValid      = "destinationCredsValid"
-	Query                      = "query"
 	QueryExecutionTime         = "queryExecutionTime"
 	StagingTableName           = "stagingTableName"
 )
+
+// TODO finish converting constants to functions to experiment
+
+type Field func(v interface{}) KeyValue
+
+type KeyValue struct {
+	k string
+	v interface{}
+}
+
+var (
+	UploadID      Field = func(v interface{}) KeyValue { return KeyValue{"uploadID", v} }
+	SourceID      Field = func(v interface{}) KeyValue { return KeyValue{"sourceID", v} }
+	DestinationID Field = func(v interface{}) KeyValue { return KeyValue{"destinationID", v} }
+	TableName     Field = func(v interface{}) KeyValue { return KeyValue{"tableName", v} }
+	Location      Field = func(v interface{}) KeyValue { return KeyValue{"location", v} }
+	Query         Field = func(v interface{}) KeyValue { return KeyValue{"query", v} }
+	Error         Field = func(v interface{}) KeyValue {
+		if err, ok := v.(error); ok {
+			return KeyValue{"error", err.Error()}
+		} else {
+			return KeyValue{"error", v}
+		}
+	}
+
+	// Generic should be used only when there is no other option and it doesn't make sense to create a new Field
+	Generic = func(k string) Field {
+		return func(v interface{}) KeyValue {
+			return KeyValue{k, v}
+		}
+	}
+)
+
+func NewLogger(component string, l logger.Logger) Logger {
+	return Logger{component: component, logger: l}
+}
+
+type Logger struct {
+	component string
+	logger    logger.Logger
+	once      sync.Once
+}
+
+func (l *Logger) init() {
+	l.once.Do(func() {
+		if l.logger == nil {
+			l.logger = logger.NOP
+		}
+	})
+}
+
+func (l *Logger) Infow(msg string, kvs ...KeyValue) {
```

Comment:
KeyValue will be good and provides flexibility and provide consistent way of naming things but they don't any meaning associated with them. Do we need a structure like following?
type Metadata struct {
 DestinationID string
 WorkspaceID string
 SourceID string
 OtherFileds map[string]string
}
something like this may be opinionated but now we know what is workspaceID and sourceID etc so we can integrate them with various tools like loki more easily.

Reply:
Don't we already know that with the approach I'm suggesting? Can you try to explain the pros and cons of your approach and how it would differ from this? I don't think I fully understand your idea.

Reply:
If we need validation of this metadata like DestinationID and WorkspaceID required etc then having something like above structure it is easy to do, also Metadata structure can vary in different situation and teams can extend the base logger with their own validation rules. May be underlying we can have generic key value logger with consistent names and on top we can have metadata structure of teams to easily define validation rules etc. Hence my approach requires work from each team to define their own metadata structure

Reply:
I can try to have a stab at this and push it in another branch to compare.

Reply:
@koladilip is [this](https://github.com/rudderlabs/rudder-server/blob/chore.whLogfield-2/warehouse/logfield/logfield.go) close enough to what you mean?

If yes I don't like these things about it:
* slower computation, the struct can become quite big and every time we log we have to check every field manually versus iterating over a slice
* maintenance, we need to manually add both a field in the struct and the related `if` vs 1 function
* we lose the ability to use the `Generic` function which allows for a more relaxed approach. down the line you simply "Find usages" of the `Generic` function and start to organize potential new fields together

What we can't easily do with the functions approach is validating multiple fields but it sounds like we don't need it (unless you have a specific example in mind) and I personally find that overkilling for a logging middleware.

Validation of single fields can still be done in the function where you define the field itself instead of being decoupled and have the definition in one place (struct attribute) and the validation down below (in the method).

Thoughts?